File: test-suite/functions/nodejs/stream-processor/handler.js
/**
 * Stream Processing Handler
 * Use case: Real-time data stream transformation and filtering
 */

exports.handler = async (event, context) => {
    const { operation = 'process', stream = [], config = {} } = event;
    
    try {
        if (operation === 'process') {
            // Process stream of events
            const results = [];
            let processed = 0;
            let filtered = 0;
            let errors = 0;
            
            for (const item of stream) {
                try {
                    // Apply filters
                    if (config.filters) {
                        let passedFilters = true;
                        
                        for (const [key, value] of Object.entries(config.filters)) {
                            if (item[key] !== value) {
                                passedFilters = false;
                                break;
                            }
                        }
                        
                        if (!passedFilters) {
                            filtered++;
                            continue;
                        }
                    }
                    
                    // Transform item
                    let transformed = { ...item };
                    
                    if (config.transformations) {
                        for (const [key, transform] of Object.entries(config.transformations)) {
                            if (transform === 'uppercase' && typeof transformed[key] === 'string') {
                                transformed[key] = transformed[key].toUpperCase();
                            } else if (transform === 'lowercase' && typeof transformed[key] === 'string') {
                                transformed[key] = transformed[key].toLowerCase();
                            } else if (transform === 'multiply' && typeof transformed[key] === 'number') {
                                transformed[key] = transformed[key] * (config.multiplier || 2);
                            }
                        }
                    }
                    
                    // Add metadata
                    transformed._processed_at = new Date().toISOString();
                    transformed._processed_by = 'stream-processor';
                    
                    results.push(transformed);
                    processed++;
                    
                } catch (err) {
                    errors++;
                }
            }
            
            return {
                statusCode: 200,
                body: JSON.stringify({
                    results,
                    statistics: {
                        total: stream.length,
                        processed,
                        filtered,
                        errors,
                        success_rate: (processed / stream.length * 100).toFixed(2) + '%'
                    }
                })
            };
        } else if (operation === 'aggregate') {
            // Aggregate stream data
            const aggregates = {};
            const groupBy = config.groupBy || 'type';
            
            for (const item of stream) {
                const key = item[groupBy] || 'unknown';
                
                if (!aggregates[key]) {
                    aggregates[key] = {
                        count: 0,
                        sum: 0,
                        values: []
                    };
                }
                
                aggregates[key].count++;
                
                if (typeof item.value === 'number') {
                    aggregates[key].sum += item.value;
                    aggregates[key].values.push(item.value);
                }
            }
            
            // Calculate averages
            for (const key in aggregates) {
                if (aggregates[key].values.length > 0) {
                    aggregates[key].average = 
                        aggregates[key].sum / aggregates[key].values.length;
                    aggregates[key].min = Math.min(...aggregates[key].values);
                    aggregates[key].max = Math.max(...aggregates[key].values);
                }
                delete aggregates[key].values; // Remove raw values
            }
            
            return {
                statusCode: 200,
                body: JSON.stringify({
                    aggregates,
                    total_items: stream.length,
                    unique_groups: Object.keys(aggregates).length
                })
            };
        } else if (operation === 'window') {
            // Windowing operations
            const windowSize = config.windowSize || 10;
            const windows = [];
            
            for (let i = 0; i < stream.length; i += windowSize) {
                const window = stream.slice(i, i + windowSize);
                const windowStats = {
                    window_id: Math.floor(i / windowSize),
                    size: window.length,
                    start_index: i,
                    end_index: Math.min(i + windowSize - 1, stream.length - 1)
                };
                
                // Calculate window statistics
                const values = window
                    .map(item => item.value)
                    .filter(v => typeof v === 'number');
                
                if (values.length > 0) {
                    windowStats.sum = values.reduce((a, b) => a + b, 0);
                    windowStats.average = windowStats.sum / values.length;
                    windowStats.min = Math.min(...values);
                    windowStats.max = Math.max(...values);
                }
                
                windows.push(windowStats);
            }
            
            return {
                statusCode: 200,
                body: JSON.stringify({
                    windows,
                    window_size: windowSize,
                    total_windows: windows.length
                })
            };
        }
        
        return {
            statusCode: 400,
            body: JSON.stringify({ error: 'Unknown operation' })
        };
        
    } catch (error) {
        return {
            statusCode: 500,
            body: JSON.stringify({ error: error.message })
        };
    }
};

File: server/tests/test-suite/functions/java/batch-processor/Handler.java
import java.util.*;
import java.util.stream.Collectors;

/**
 * Batch Processing Handler
 * Use case: High-volume data processing with parallel execution
 */
public class Handler {
    
    public Map<String, Object> handler(Map<String, Object> event, Map<String, Object> context) {
        String operation = (String) event.getOrDefault("operation", "process");
        List<Map<String, Object>> batch = (List<Map<String, Object>>) 
            event.getOrDefault("batch", new ArrayList<>());
        Map<String, Object> config = (Map<String, Object>) 
            event.getOrDefault("config", new HashMap<>());
        
        try {
            switch (operation) {
                case "process":
                    return processBatch(batch, config);
                case "validate":
                    return validateBatch(batch, config);
                case "transform":
                    return transformBatch(batch, config);
                case "aggregate":
                    return aggregateBatch(batch, config);
                default:
                    return createErrorResponse(400, "Unknown operation: " + operation);
            }
        } catch (Exception e) {
            return createErrorResponse(500, "Processing error: " + e.getMessage());
        }
    }
    
    private Map<String, Object> processBatch(List<Map<String, Object>> batch, 
                                              Map<String, Object> config) {
        long startTime = System.currentTimeMillis();
        
        List<Map<String, Object>> processed = new ArrayList<>();
        List<Map<String, Object>> failed = new ArrayList<>();
        
        for (Map<String, Object> item : batch) {
            try {
                // Simulate processing
                Map<String, Object> processedItem = new HashMap<>(item);
                processedItem.put("processed_at", new Date().toString());
                processedItem.put("status", "completed");
                
                // Apply business logic
                if (item.containsKey("value")) {
                    Object value = item.get("value");
                    if (value instanceof Number) {
                        double numValue = ((Number) value).doubleValue();
                        processedItem.put("squared", numValue * numValue);
                        processedItem.put("doubled", numValue * 2);
                    }
                }
                
                processed.add(processedItem);
            } catch (Exception e) {
                Map<String, Object> errorItem = new HashMap<>(item);
                errorItem.put("error", e.getMessage());
                failed.add(errorItem);
            }
        }
        
        long duration = System.currentTimeMillis() - startTime;
        
        Map<String, Object> result = new HashMap<>();
        result.put("processed", processed);
        result.put("failed", failed);
        result.put("statistics", Map.of(
            "total", batch.size(),
            "successful", processed.size(),
            "failed", failed.size(),
            "duration_ms", duration,
            "throughput", batch.size() * 1000.0 / duration
        ));
        
        return createResponse(200, result);
    }
    
    private Map<String, Object> validateBatch(List<Map<String, Object>> batch,
                                               Map<String, Object> config) {
        List<String> requiredFields = (List<String>) 
            config.getOrDefault("required_fields", new ArrayList<>());
        
        List<Map<String, Object>> valid = new ArrayList<>();
        List<Map<String, Object>> invalid = new ArrayList<>();
        
        for (Map<String, Object> item : batch) {
            List<String> missingFields = new ArrayList<>();
            
            for (String field : requiredFields) {
                if (!item.containsKey(field) || item.get(field) == null) {
                    missingFields.add(field);
                }
            }
            
            if (missingFields.isEmpty()) {
                valid.add(item);
            } else {
                Map<String, Object> invalidItem = new HashMap<>(item);
                invalidItem.put("missing_fields", missingFields);
                invalid.add(invalidItem);
            }
        }
        
        Map<String, Object> result = new HashMap<>();
        result.put("valid", valid);
        result.put("invalid", invalid);
        result.put("validation_rate", (double) valid.size() / batch.size() * 100);
        
        return createResponse(200, result);
    }
    
    private Map<String, Object> transformBatch(List<Map<String, Object>> batch,
                                                Map<String, Object> config) {
        String transformType = (String) config.getOrDefault("type", "uppercase");
        
        List<Map<String, Object>> transformed = batch.stream()
            .map(item -> {
                Map<String, Object> newItem = new HashMap<>(item);
                
                if ("uppercase".equals(transformType)) {
                    item.forEach((key, value) -> {
                        if (value instanceof String) {
                            newItem.put(key, ((String) value).toUpperCase());
                        }
                    });
                } else if ("lowercase".equals(transformType)) {
                    item.forEach((key, value) -> {
                        if (value instanceof String) {
                            newItem.put(key, ((String) value).toLowerCase());
                        }
                    });
                }
                
                return newItem;
            })
            .collect(Collectors.toList());
        
        return createResponse(200, Map.of("transformed", transformed));
    }
    
    private Map<String, Object> aggregateBatch(List<Map<String, Object>> batch,
                                                Map<String, Object> config) {
        String groupBy = (String) config.getOrDefault("group_by", "type");
        
        Map<String, List<Map<String, Object>>> groups = batch.stream()
            .collect(Collectors.groupingBy(
                item -> String.valueOf(item.getOrDefault(groupBy, "unknown"))
            ));
        
        Map<String, Map<String, Object>> aggregates = new HashMap<>();
        
        for (Map.Entry<String, List<Map<String, Object>>> entry : groups.entrySet()) {
            String key = entry.getKey();
            List<Map<String, Object>> items = entry.getValue();
            
            double sum = items.stream()
                .filter(item -> item.containsKey("value"))
                .mapToDouble(item -> ((Number) item.get("value")).doubleValue())
                .sum();
            
            aggregates.put(key, Map.of(
                "count", items.size(),
                "sum", sum,
                "average", items.size() > 0 ? sum / items.size() : 0
            ));
        }
        
        return createResponse(200, Map.of(
            "aggregates", aggregates,
            "total_groups", groups.size()
        ));
    }
    
    private Map<String, Object> createResponse(int statusCode, Object body) {
        Map<String, Object> response = new HashMap<>();
        response.put("statusCode", statusCode);
        response.put("body", body);
        return response;
    }
    
    private Map<String, Object> createErrorResponse(int statusCode, String message) {
        return createResponse(statusCode, Map.of("error", message));
    }
}
